Add integration test

benhyland opened the pull request:
Adds an integration test that runs most of the app code.

Main purpose is to detect basic code errors our static analysis doesn't pick up.
Secondary purpose is to check that the program does something vaguely expected.
We run all checks, and import a tiny bit of data from stubbed http endpoints.
We check that real csv files have been written.

This PR takes the form of several commits gradually fleshing out the test, interspersed with minor functionality additions to the program to make it easier to configure for the test.
The commits are probably best reviewed individually.

Commit:
Run all checks, when check names are not specified

Previously, check names were mandatory arguments.
Now we detect when no check names were passed, and run all checks.
This makes it slightly easier to maintain the integration test we will write.

Commit:
Add integration test

Use webmock services to run (roughly) the whole program in a spec.
Just the test outline for now, need to stub out a real data import and
examine the csv output.

malcolmbaig commented on lib/checks/report/report.rb:
```diff
@@ -0,0 +1,28 @@
+module Checks
+  module Report
+    class Report
```
Reckon you're better off moving report.rb into lib and dropping this slightly odd namespacing.

benhyland replied:
Maybe. I hope there is a better alternative.
I don't really want anything other than checker.rb and check_runner.rb in lib (the progress_reporter thing is its own concern and doesn't really belong there).
The Report class is a common feature of checks and belongs somewhere under the checks dir and in the Checks namespace.
I moved it from being a direct entry in the check dir as this makes loading all checks by default easier.

Davidslv replied:
I assume you referring to the following line here? https://github.com/alphagov/finding-things-migration-checker/pull/18/files#diff-e496d5ba7c98bf92be59d9f4819dddbbR37

I don't know of a better way, but what if you added an `unless` statement, `require file unless file == 'report'` or something along those lines?

benhyland replied:
Yep that's right.
I think special casing like that should be avoided as it is very fragile to future change (even more so than doing the dynamic loading in the first place)

MatMoore replied:
I was actually wondering if we could move the report creation out of the checks. It seems fairly standard but I haven't looked into it yet.

eg could we just get the headers and results from the check object, and have the runner create a report from that (and the whitelist)?

benhyland replied:
The purpose of the report is to encapsulate the stuff the checks need to return.
We could certainly move the whitelisting out, but not the report creation.

... I just understood that you might be suggesting we keep some mutable state on each check.
I think this would not be a good idea - I prefer immutability for reasons too numerous for this discussion, wherever performance is not at issue.

MatMoore replied:
The alternative I was thinking of is something like -

``` ruby
class SomeCheck
  HEADERS = %w(...)

  def run(...)
    ...
    results
 end
end
```

With the Report being constructed by the caller. I'm fine with addressing this in a separate PR (or not at all) though.

benhyland replied:
That would be OK. I think we should address it in a later PR (when we're next in the area) as well.

Commit:
Allow specifying the csv output directory

This is helpful so we can clean up the integration test output and identify
build artefacts in Jenkins more easily.

Commit:
Allow supressing progress reporting

Makes for cleaner integration test output

Davidslv commented on lib/check_runner.rb:
```diff
@@ -1,24 +1,26 @@
     whitelist = Whitelist.load(whitelist_file)
 
-    progress_reporter = ProgressReporter.new
+    @progress_reporter = suppress_progress ? ProgressReporter.noop : ProgressReporter.new
```
What you think of this be part of the argument received by `ProgressReporter` constructor?

`ProgressReporter.new(supress_progress: suppress_progress)`

benhyland replied:
Usually it is a good idea to do all the config and wiring in one place, but I don't feel too strongly about it.

Commit:
Stub a data item in the Rummager and Publishing API public endpoints

We're now integration testing a reasonable amount -
further testing should probably focus on individual checks,
or the execution model. Enough for now.

MatMoore commented on spec/integration/check_runner_spec.rb:
```diff
@@ -0,0 +1,167 @@
+  end
+
+  def check_csv_content
+    expect(read_csv('LinksMissingFromRummager.csv')).to eq("link_type,link_content_id,content_id,publishing_app,format\norganisations,42,1,app1,format1\n")
```
Is it possible to stub out the filesystem in this test? Have it write to some [StringIO](http://ruby-doc.org/stdlib-2.3.0/libdoc/stringio/rdoc/StringIO.html) objects or something?

benhyland replied:
It is possible, but I'd rather an integration test test the real filesystem if it is not too onerous to do so.

MatMoore commented on spec/integration/check_runner_spec.rb:
```diff
@@ -0,0 +1,167 @@
+    )
+  end
+
+  def stub_publishing_api_content_request(request_last_seen_content_id:, response_last_seen_content_id:, response_results:)
```
There is a central stub for this, `publishing_api_has_grouped_content_and_links` - but it's not actually merged yet. Maybe we can simplify this when we do the 2nd pass of the publishing api request. https://github.com/alphagov/gds-api-adapters/pull/510/files

Commit:
Allow whitelist to correctly handle an empty file

Use an empty whitelist in the integration test.
Whitelist is reasonably well unit tested, and we aren't trying to
observe it in the integration test.

Commit:
Use the 'run all checks by default' behaviour in the jenkins job

So we don't have to remember something else when adding checks in future

Commit:
Remove obsolete comment
